test/integration/etcd_storage_path_test.go

```diff
@@ -0,0 +1,599 @@
+}{
+	gvr("servicecatalog.k8s.io", "v1beta1", "clusterservicebrokers"): {
+		stub:             `{"metadata": {"name": "broker1"}, "spec": {"url": "https://broker1.com"}}`,
+		expectedEtcdPath: "/servicecatalog.k8s.io/clusterservicebrokers/broker1",
+	},
+	gvr("servicecatalog.k8s.io", "v1beta1", "clusterserviceclasses"): {
+		stub:             `{"metadata": {"name": "class1"}, "spec": {"clusterServiceBrokerName": "broker1", "externalName": "class-name1", "externalID": "class1", "description": "desc"}}`,
+		expectedEtcdPath: "/servicecatalog.k8s.io/clusterserviceclasses/class1",
+	},
+	gvr("servicecatalog.k8s.io", "v1beta1", "clusterserviceplans"): {
+		stub:             `{"metadata": {"name": "plan1"}, "spec": {"clusterServiceBrokerName": "broker1", "externalName": "plan-name1", "externalID": "plan1", "description": "desc", "clusterServiceClassRef": {"Name": "class1"}}}`,
+		expectedEtcdPath: "/servicecatalog.k8s.io/clusterserviceplans/plan1",
+	},
+	gvr("servicecatalog.k8s.io", "v1beta1", "serviceinstances"): {
+		stub:             `{"metadata": {"namespace": "etcdstoragepathtestnamespace", "name": "instance1"}, "spec": {"clusterServiceClassExternalName": "class1", "clusterServicePlanExternalName": "plan1", "externalID": "instance1"}}`,
+		expectedEtcdPath: "/servicecatalog.k8s.io/serviceinstances/etcdstoragepathtestnamespace/instance1",
+	},
+	gvr("servicecatalog.k8s.io", "v1beta1", "servicebindings"): {
+		stub:             `{"metadata": {"namespace": "etcdstoragepathtestnamespace", "name": "binding1"}, "spec": {"instanceRef": {"name": "instance1"}, "externalID": "binding1"}}`,
+		expectedEtcdPath: "/servicecatalog.k8s.io/servicebindings/etcdstoragepathtestnamespace/binding1",
+	},
+}
+
+// Be very careful when whitelisting an object as ephemeral.
+// Doing so removes the safety we gain from this test by skipping that object.
+var ephemeralWhiteList = createEphemeralWhiteList()
+
+// Only add kinds to this list when there is no way to create the object
+var kindWhiteList = sets.NewString(
+	// k8s.io/kubernetes/pkg/api/v1
+	"DeleteOptions",
+	"ExportOptions",
+	"ListOptions",
+	"GetOptions",
+	"APIGroup",
+	"APIVersions",
+	// --
+
+	// k8s.io/kubernetes/pkg/watch/versioned
+	"WatchEvent",
+	// --
+
+	// k8s.io/kubernetes/pkg/api/unversioned
+	"Status",
+	// --
+)
+
+// namespace used for all tests, do not change this
+const testEtcdStorageNamespace = "etcdstoragepathtestnamespace"
+
+// TestEtcdStoragePath tests to make sure that all objects are stored in an expected location in etcd.
+// It will start failing when a new type is added to ensure that all future types are added to this test.
+// It will also fail when a type gets moved to a different location. Be very careful in this situation because
+// it essentially means that you will be break old clusters unless you create some migration path for the old data.
```

Comment:
I'd love to know anything about having a migration path for old data. We should have one example one in place for future. As it is now, it looks like we just break stuff and say "too bad", start over.

Reply:
I'm not entirely sure what you are looking for here. This test will fail when the strategy for a resource is changed to store it in a different location. That is good. The warning is telling you that you shouldn't just change the location in this test to match the new location without also providing a migration path from the old location to the new location. It is well beyond the scope of this PR to come up with migration plans for some future potential location change.